### ISSUE TITLE
file path was truncated by character '?' when using S3 UFS

### ISSUE THREAD

**Issue by wwq2333** [Alluxio/alluxio#17792]
**Alluxio Version:**
master-2.x

**Describe the bug**
we use s3 as alluxio UFS;
when then file path (s3 prefix) has character '?',  use `alluxio fs ls /path` or `shell ls /path` via fuse, the path was truncated by character '?'

<img width="1161" alt="image" src="https://github.com/Alluxio/alluxio/assets/12540485/054d42df-9517-4f37-b37e-afb754dc0dd8">

**Expected behavior**
can index the file via fuse correctly

**Comment by elega** [comment]
@HelloHorizon Hi can we get someone take a look as this is a high priority issue that blocks their production workflow.

**Comment by elega** [comment]
Alluxio doesn't quite support files that contains '?' in its name. 

![3d8a6db1c6d5819f190e7217ca8e04a](https://github.com/Alluxio/alluxio/assets/6771554/4b5cb4eb-8590-44a6-b70d-dba570b0835f)

Here we filtered files with '?'
https://github.com/Alluxio/alluxio/blob/main/dora/core/common/src/main/java/alluxio/underfs/UnderFileSystemWithLogging.java#L861

You can try get rid of the invalid path filtering thing but i doubt more errors will come up as we have made the assumption that such file name is not valid everywhere.

**Comment by jja725** [comment]
@yuzhu @jiacheliu3 Do you know why we have this limitation in alluxio uri? And is this limitation still valid in Dora(I assume no)?

**Comment by jja725** [comment]
@jiacheliu3 And I believe during the process of adopting UFSURi (instead of AlluxioURI), we should be able to resolve this problem?

**Comment by yyongycy** [comment]
Can we know the business justification? Is it a-must-have or by chance?

**Comment by Jackson-Wang-7** [comment]
S3 don't support the object key with '?' directly. you can check with this link here. https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

So we follow the limitation to not support the file name includes '?'. If you have the requirement, you can encode the '?' in the filename and then use it.

**Comment by humengyu2012** [comment]
Because object storage uses the HTTP protocol, the '?' character may conflict with HTTP queries.

**Comment by jja725** [comment]
@yyongycy it would potentially be a blocker since we can see the file in s3 but can't do anything with it
@jiacheliu3 Since we totally use the UFS path as the key right now I think we should be able to support the special characters which the UFS supports

**Comment by jiacheliu3** [comment]
As @Jackson-Wang-7 pointed out, some special characters are hard to handle, especially hard if you have to serve multiple APIs. That was why s3 discourages people from using it, in https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html

However, I agree with @jja725 that the Alluxio layer doesn't have to put extra filters and just reject "?".

I wasn't at Alluxio when that filter @elega pointed out was added, but here's my guess: `AlluxioURI` wants to be conform to the definition of `URI`. A URI follows the pattern of `<scheme>://<authority>/<path>?<query>` so the "?" belongs to the query part. So the path does not allow "?"
https://github.com/Alluxio/alluxio/blob/ae7439e7f073fd00a09b32984465f4a64773a4ec/dora/core/common/src/main/java/alluxio/AlluxioURI.java#L498
I don't know if there was a specific use case for the `<query>` part in URI. I'm not aware of one.

Regarding what @humengyu2012 raised, I think path with "?" is achievable even though Alluxio may receive and send HTTP requests. Correct me if I'm wrong on the below process I imagine:
```
Assume we want to access Alluxio path /s3/bucket/a?b. The object is in s3://bucket/a?b

The user uses Alluxio RESTful API. Then the user has to encode the path to /s3/bucket/a%3Fb, with "?" -> "%3F". The final HTTP request may look like "https://<alluxio-host>:<proxy port>/..../s3/bucket/a%3Fb?query=..."

Alluxio gets the request and can decode `/s3/bucket/a?b` following the same rule.

Alluxio sends GetObject(bucket, "a?b") to S3.
```

**Comment by jja725** [comment]
@jiacheliu3 yes I totally agree about the HTTP request handling, @beinan @JiamingMai FYi
